### DISCUSSION TITLE
Use deployer to download a file from a server

### DISCUSSION THREAD

**Discussion by alexander-schranz** [deployphp/deployer#3610] [Help needed]
I want to add a task to download a file from the server. So connect to the `current` directory `run` a command which generates a `zip` and then download that via deployer to the local system. So it is totally another process then deployment, but thought maybe this kind of things is also possible via deployer to do it reverse get files from the server?

My usecase is that I want to generate a backup and download that to my local system. As deployer already have kind of possibility to connect to the server I thought maybe that could be possible to instead of upload a file to download something.

Is this with a custom task possible?

**Comment by antonmedv** [comment]
Yes, you can indeed use Deployer to accomplish this task. Deployer's capabilities are not restricted to only deploying your application, but also managing and executing arbitrary tasks on your servers. You can define a custom task that generates a backup of your desired files or directories and then downloads it to your local machine. Here's an example of how you could do it:

```php
namespace Deployer;

desc('Creating and downloading backup');
task('backup:download', function () {
    $host = Context::get()->getHost();
    $user = Context::get()->getUser();
    $path = '~/backup.zip'; // Define your backup path here.

    // Generate the zip file on the server
    run("zip -r {$path} ."); // This will zip the current directory.

    // Download the file to the local machine
    download($path, __DIR__ . "/{$host}.zip");
});

after('deploy', 'backup:download');
```

This task `backup:download` first generates a zip file of the current directory on the server, then it uses the `download` function of Deployer to download that file to the local machine.

Please replace `'~'` with the directory you want to zip and backup on your server. Also replace `'.'` with the files/directories you want to include in the zip file.

You can run the task manually with the `dep backup:download` command or automatically after the `deploy` task as in the example.

Please note that the `download` function will download the file to the directory where you're running your Deployer script from. If you want to specify a different directory, you should adjust the second parameter of the `download` function accordingly.

Also, please ensure that your server has the `zip` utility installed and available in its `PATH`, as it's being used to generate the zip file. If it's not available or you prefer a different archiving tool, you would have to adjust the command accordingly.
